crates/uv-installer/src/compile.rs

```diff
@@ -0,0 +1,207 @@
+        .stdin(Stdio::piped())
+        .stdout(Stdio::piped())
+        .stderr(Stdio::inherit())
+        .current_dir(dir)
+        // Otherwise stdout is buffered and we'll wait forever for a response
+        .env("PYTHONUNBUFFERED", "1")
+        .spawn()
+        .map_err(CompileError::PythonSubcommand)?;
+
+    // https://stackoverflow.com/questions/49218599/write-to-child-process-stdin-in-rust/49597789#comment120223107_49597789
+    // Unbuffered, we need to write immediately or the python process will get stuck waiting
+    let child_stdin = bytecode_compiler
+        .stdin
+        .take()
+        .expect("Child must have stdin");
+    let mut child_stdout = BufReader::new(
+        bytecode_compiler
+            .stdout
+            .take()
+            .expect("Child must have stdout"),
+    );
+
+    let result = worker_main_loop(receiver, child_stdin, &mut child_stdout).await;
+    // Avoid `/.venv/bin/python: can't open file` on stderr after success because the tempdir was
+    // dropped before the Python process even started.
+    let _ = bytecode_compiler.kill().await;
+    result
+}
+
+/// We use stdin/stdout as a sort of bounded channel. We write one path to stdin, then wait until
+/// we get the same path back from stdout. This way we ensure one worker is only working on one
+/// piece of work at the same time.
+async fn worker_main_loop(
+    receiver: Receiver<PathBuf>,
+    mut child_stdin: ChildStdin,
+    child_stdout: &mut BufReader<ChildStdout>,
+) -> Result<(), CompileError> {
+    let mut out_line = String::new();
+    while let Ok(source_file) = receiver.recv().await {
+        // Luckily, LF alone works on windows too
+        let bytes = format!("{}\n", source_file.display()).into_bytes();
+        // Ensure we don't get stuck here because some pipe ran full. The Python process isn't doing
+        // anything at this point except waiting for stdin, so this should be immediate.
+        let timeout = Duration::from_secs(1);
+        match tokio::time::timeout(timeout, child_stdin.write_all(&bytes)).await {
+            // Timeout
+            Err(_) => {
+                receiver.close();
+                return Err(CompileError::Timeout(timeout));
+            }
+            // Write error
+            Ok(Err(err)) => return Err(CompileError::ChildStdin(err)),
+            Ok(Ok(())) => {}
+        }
+
+        out_line.clear();
+        child_stdout
+            .read_line(&mut out_line)
+            .await
+            .map_err(CompileError::ChildStdout)?;
+
+        // This is a sanity check, if we don't get the path back something has gone wrong, e.g.
+        // we're not actually running a python interpreter.
+        let actual = out_line.trim_end_matches(['\n', '\r']);
+        let expected = source_file.display().to_string();
+        if expected != actual {
+            return Err(CompileError::WrongPath(expected, actual.to_string()));
+        }
+    }
+    Ok(())
+}
```

Comment:
So looking at this code holistically, I feel like we might be able to simplify the timeout/cancellation logic here.

The way I would go about this is to think about the workers themselves controlling how long they're willing to wait for the subprocess to complete its task. You pick a timeout at that granularity. Then right before to start communicating with the subprocess, spawk a task that will kill the subprocess if it hasn't completed after N seconds. You should be able to implement this with `tokio::select!`.

If you do that, then I think you should be able to get rid of the timeouts almost everywhere else. It absolves of you needing to worry about writing to stdin taking too long, because the worker will kill the subprocess if it's truly blocked. When the subprocess gets killed, your code that is blocked on stdin should become unblocked immediately and everything should gracefully terminate.

Reply:
I've simplified the timeouts in a different way: Only the caller has a timeout for workers (or the channel) getting stuck, one when sending (due to backpressure) and one when waiting for the workers to join. This way it doesn't matter whether the python process is broken, there's a bug in the worker implementation or the tokio threadpool or the channel broke due to a panic, each case trips that timeout.

Placing it at the send might be an unusual positioning, but i think it's the best place to catch stuck workers.

Reply:
Note that either case is almost certainly a bug, either a real bug in uv or a quirk python interpreter we need to special case.